### PR TITLE
feat: better unique identifier support

### PR DESCRIPTION
**Pull request by vmx**
Using the UUID as unique identifier is preferred, though it is not always
available. Hence there is a fallback to the PCI ID, which is almost always
available on all supported devices. The only case it is not available is
on MacBooks with an AMD graphics card. Those machines are used for
development, but not in production.

BREAKING CHANGE: Bus ID is replaced with PCI ID

Prior to this change only the so-called "Bus ID" was used. Instead now use
a combination of the device and the Bus Id, which should provide more
uniqueness.

Also the `GPUSelector` is replaced by with additional functions on
`Device` which provide the same functionality.

### CODE REVIEW

**Comment by dignifiedquire** [comment]
there is a lot of parsing and formating in this PR, can you please add unit tests for those

**Comment by vmx** [comment]
Those tests are here: https://github.com/filecoin-project/rust-gpu-tools/blob/43dbc5485ef45f7a3dd1002dbedb8d73f6d03ded/src/device.rs#L409-L470 (which is a later version, which doesn't have a PR yet). But I can backport them if needed.

**Comment by vmx** [comment]
Now with tests :)

**Comment by dignifiedquire** [comment]
left some comments on the other PR because I was confused 😅

**Comment by vmx** [comment]
Oh wow, making sure that the fake PCI-IDs don't collide with actual ones was tougher then I thought. Though you now also get a `filter_map()` as requested elsewhere.

**Comment by vmx** [comment]
I've addressed the comments from the other PR (#40). Sorry for the confusion, something went wrong with the branches somehow. Now everything should be in place again :)

File: src/opencl/utils.rs [hunk 1]
```diff
@@ -1,19 +1,45 @@
...
 
-fn get_bus_id(d: &opencl3::device::Device) -> Result<u32, GPUError> {
-    let vendor = Vendor::try_from(d.vendor()?.as_str())?;
-    match vendor {
-        Vendor::Amd => d.pci_bus_id_amd().map_err(Into::into),
-        Vendor::Nvidia => d.pci_bus_id_nv().map_err(Into::into),
-    }
+/// The PCI-ID is the combination of the PCI Bus ID and PCI Device ID.
+///
+/// It is the first two identifiers of e.g. `lcpci`:
```
Comment (dignifiedquire) [inline src/opencl/utils.rs:12]: typo

Comment (vmx) [reply src/opencl/utils.rs]: There were more places then I thought with this typo ;)

File: src/opencl/mod.rs [hunk 2]
```diff
@@ -283,7 +390,7 @@ impl Program {
         offset: usize,
         data: &[T],
     ) -> GPUResult<()> {
-        assert!(offset + data.len() <= buffer.length, "Buffer is too small.");
```
Comment (dignifiedquire) [inline src/opencl/mod.rs:286]: why you delete this?

File: src/opencl/utils.rs [hunk 3]
```diff
@@ -1,19 +1,45 @@
...
+        }
+    };
+    Ok(id.into())
+}
+
+fn get_uuid(device: &opencl3::device::Device) -> GPUResult<DeviceUuid> {
+    let uuid: [u8; CL_UUID_SIZE_KHR] = device
+        .uuid_khr()?
+        .try_into()
+        .expect("opencl3 returned an invalid UUID");
```
Comment (dignifiedquire) [inline src/opencl/utils.rs:41]: this should include the invalid id, other this will be impossible to debug

File: src/opencl/mod.rs (hunk 2 above)
Comment (vmx) [reply src/opencl/mod.rs:283]: Rebasing failure.

**Review by vmx — COMMENTED** [review]
I've pushed the requested changes, I'll rebase once those are reviewed again, to make the review easier.

**Comment by vmx** [comment]
The new code doesn't work :-/ Back to the drawing board.

**Comment by vmx** [comment]
Issue fixed, ready for review. The `opencl3` API isn't ideal, it's easy to get 0s and 1s wrong. I opened an issue for that: https://github.com/kenba/opencl3/issues/21.

**Comment by vmx** [comment]
A rebased version of this code was tested on a machine that doesn't report a PCI-ID and it worked as expected.

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**